add spectral normalization [pytorch]

crcrpar opened the pull request:
related to #5027.
This PR aims at implementing Spectral Normalization in a way similar to `torch.nn.utils.weight_norm`.

crcrpar committed:
initial commit for spectral norm

crcrpar committed:
fix comment

crcrpar committed:
edit rst

soumith commented:
@pytorchbot retest this please

crcrpar commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,117 @@
+    return v / denom
+
+
+class SpectralNorm(object):
```
This normalizes weights of layers, not outputs of layers.
So, I think it's preferable to implement like `WeightNorm` for flexibility.

crcrpar replied:
@ezyang What do you think about this?

crcrpar committed:
fix doc

crcrpar committed:
remove redundant empty line

SsnL commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,117 @@
+         \mathbf{W} &= \dfrac{\mathbf{W}}{\sigma(\mathbf{W})} \\
+         \sigma(\mathbf{W}) &= \max_{\mathbf{h}: \mathbf{h} \ne 0} \dfrac{\|\mathbf{W} \mathbf{h}\|_2}{\|\mathbf{h}\|_2}
+
+    Spectral normalization stabilize the training of discriminators(critics)
```
nit: stabilizes, space before `(`

crcrpar replied:
thanks

SsnL commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,117 @@
+    Spectral normalization stabilize the training of discriminators(critics)
+    in GANs by rescaling the weight tensor by spectral norm "sigma" of the
+    weight matrix calculated by power iteration method. If the dimension of the
+    weight tensor is greater than 2, reahaped to 2D in power iteration method
```
nit reshaped

crcrpar replied:
thank you

crcrpar committed:
fix nit mistakes in doc

fmassa commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,117 @@
+from torch.nn.parameter import Parameter
+
+
+def l2normalize(v, eps=1e-12):
```
Can't you use instead [`torch.nn.functional.normalize`](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2004)?

crcrpar replied:
I mimic the author's [implementation](https://github.com/pfnet-research/sngan_projection/blob/master/source/functions/max_sv.py#L5) where the denominator is `norm + eps`, not `max(norm, eps)`.

I'm not sure which is suitable so if you think `F.normalize` better, I'll replace.

fmassa replied:
I think there is no real difference actually. This is there just to avoid division by zero.

crcrpar commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,117 @@
+    def compute_weight(self, module):
+        weight = module._parameters[self.name + '_org']
+        u = module._buffers[self.name + '_u']
+        height, _cuda = weight.size(0), weight.is_cuda
```
This is my fault and I'll remove it.

crcrpar committed:
replace l2normalize with F.normalize

SsnL commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,112 @@
+
+    Spectral normalization stabilizes the training of discriminators (critics)
+    in GANs by rescaling the weight tensor by spectral norm "sigma" of the
+    weight matrix calculated by power iteration method. If the dimension of the
```
nit: this sentence has three `by`s chained together. How about 
```
Spectral normalization stabilizes the training of discriminators (critics)
in GANs by rescaling the weight tensor with spectral norm :math:`\sigma` of the 
weight matrix calculated using power iteration method.
```

crcrpar commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -21,11 +16,11 @@ def __init__(self, name='weight', n_power_iterations=1, eps=1e-12):
         for _ in range(self.n_power_iterations):
-            v = l2normalize(torch.matmul(weight_mat.t(), u), self.eps)
-            u = l2normalize(torch.matmul(weight_mat, v), self.eps)
+            v = normalize(torch.matmul(weight_mat.t(), u), dim=0, eps=self.eps)
```
Yes, `torch.matmul(weight_mat.t(), u)` returns 1d.
Since the arguments of `torch.nn.functional.normalize` are `input, p=2, dim=1, eps=1e-12`, I explicitly set `dim=0`.
Or, should I use `F.normalize`?

fmassa replied:
using `torch.nn.functional.normalize` is fine then. I just wanted to make sure that we were not changing the behavior. Thanks!

crcrpar committed:
fix chained `by`

ezyang commented:
@fmassa @SsnL is this ok to merge now?

SsnL commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,112 @@
+    Spectral normalization stabilizes the training of discriminators (critics)
+    in GANs by rescaling the weight tensor with spectral norm :math:`\sigma` of
+    the weight matrix calculated using power iteration method. If the dimension
+    of the weight tensor is greater than 2, reshaped to 2D in power iteration method
```
nit: `it is reshaped`

crcrpar replied:
thanks

SsnL commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,112 @@
+    spectral norm and rescales weight before every :meth:`~Module.forward`
+    call.
+
+    See https://arxiv.org/abs/1802.05957
```
I know that this is how weight_norm doc is. But could you write this as:
```
See `Spectral Normalization for Generative Adversarial Networks`_ .

.. _`Spectral Normalization for Generative Adversarial Networks`: https://arxiv.org/abs/1802.05957
```
? We should update that as well.

crcrpar replied:
sure

SsnL commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,112 @@
+    r"""Removes the spectral normalization reparameterization from a module.
+
+    Args:
+        module (nn.Modue): containing module
```
nit: `Module`

SsnL commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,112 @@
+        weight = module._parameters[name]
+        height = weight.size(0)
+
+        u = normalize(weight.data.new(height).normal_(0, 1), dim=0, eps=fn.eps)
```
nit: `.data` is not needed, and please use `new_empty` instead of `new`.

crcrpar replied:
I didn't know that method, thank you.

SsnL commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,112 @@
+         \sigma(\mathbf{W}) &= \max_{\mathbf{h}: \mathbf{h} \ne 0} \dfrac{\|\mathbf{W} \mathbf{h}\|_2}{\|\mathbf{h}\|_2}
+
+    Spectral normalization stabilizes the training of discriminators (critics)
+    in GANs by rescaling the weight tensor with spectral norm :math:`\sigma` of
```
I'd suggest writing it out as `Generative Adversarial Networks (GANs)`.

crcrpar replied:
OK, I'll update.

SsnL commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,112 @@
+        u = module._buffers[self.name + '_u']
+        height = weight.size(0)
+        weight_mat = weight.view(height, -1)
+        for _ in range(self.n_power_iterations):
```
Can you add a comment like:
```
Spectral norm of weight equals to `u^T W v`, where `u` and `v` are the first left and right singular vectors.
This power iteration produces approximations of `u` and `v`.
```

SsnL commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,112 @@
+        name (str, optional): name of weight parameter
+        n_power_iterations (int, optional): number of power iterations to
+            calculate spectal norm
+        eps (float, optional): epsilon for numerical stability
```
nit: please be specific, like `... for numerical stability in calculating norms`.

crcrpar replied:
I got it.

ssnl reviewed (commented):
Code looks solid. There are still some minor nits around doc and comments. I think this is good to go once those are fixed.

crcrpar committed:
fix docs

fix typos
add comments related to power iteration and epsilon
update link to the paper
make some comments specific

ssnl reviewed (commented):
LGTM except the typo

crcrpar committed:
fix typo

crcrpar commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -19,6 +19,9 @@ def compute_weight(self, module):
         weight_mat = weight.view(height, -1)
         for _ in range(self.n_power_iterations):
+            # Spectral norm of weight equals to `u^T W v`, where `u` and `v`
+            # are the first left and right singular vecors.
```
I'm sorry.

ssnl reviewed (approved):
LGTM. @fmassa , do you want to take an extra look?

SsnL commented:
Thank you! @crcrpar

crcrpar commented:
Thank you for your very kind reviews and suggestions!

ferrine commented on torch/nn/utils/spectral_norm.py:
```diff
@@ -0,0 +1,118 @@
+            u = normalize(torch.matmul(weight_mat, v), dim=0, eps=self.eps)
+
+        sigma = torch.dot(u, torch.matmul(weight_mat, v))
+        weight.data /= sigma
```
Why is it made non differentiable?

SsnL replied:
A couple bugs of this initial implementation is fixed before release, including making it differentials.
